crates/conjure_core/src/ast.rs

```diff
@@ -102,3 +102,71 @@ pub enum Expression {
     SumLeq(Vec<Expression>, Box<Expression>),
     Ineq(Box<Expression>, Box<Expression>, Box<Expression>),
 }
+
+impl Expression {
```

Comment:
this is the kind of thing a lens/optics library can automate I am guessing (without looking into the rust libs)

Reply:
Shouldn't stop merging this now, but definitely would make this more scalable.

https://rust-unofficial.github.io/patterns/functional/optics.html
https://github.com/TOETOE55/lens-rs/blob/master/guide.md

Reply:
`lens-rs` seems quite nice:

```
#[derive(Copy, Clone, Debug, Review, Prism)]
enum Either<L, R> {
    #[optic]
    Left(L), // generate optics::Left
    #[optic]
    Right(R), // generate optics::Right
}

#[derive(Copy, Clone, Debug, Lens)]
struct Tuple<A, B>(#[optic] A, #[optic] B); 
// use optics::_0 or optics::_1 to access it

#[derive(Copy, Clone, Debug, Lens)]
struct Foo<A, B> {
    #[optic]
    a: A, // generate optics::a
    #[optic]
    b: B, // generate optics::b
}

#[derive(Clone, Debug, Lens)]
struct Bar {
    #[optic]
    a: String, // generate optics::a, same as above
    #[optic]
    c: i32,    // generate optics::c
}
```

Reply:
Querying , modification, construction works through a macro:

```rs
let y: _ = optic!(Ok.Some.Err).review((1, 2));
// Ok(Some(Err((1, 2))))
```

Reply:
If the exact syntax isn't too our liking, we could construct something similar with macros without too much effort.